### ISSUE TITLE
[Image processing] Exif-coordinates should be considered when uploading from app

### ISSUE THREAD

**Issue by de-vries** [mapillary/mapillary_issues#1070]
Two days ago I shot a sequence with the app. For the first part of it the gps wasn't working so I later manually added the coordinates, then I used the app to upload the photos. All photo's got uploaded without errors, but only the photos I took with working gps appeared on the map. The coords I  manually added got ignored and those photos were given the coordinates [0,0](http://www.openstreetmap.org/?mlat=0.000&mlon=-0.000#map=6/0.000/-0.000) by Mapillary.

*I already did a manual upload and requested the old sequence to be hiden/removed, so there is no need to look in to this specific case any more.*

It seems that when you process photos from the app you only considerer coordinates in the Exif-description tag and not those in the Exif-lat/long tags.
**I recommend that you not only start considering the Exif-lat/long tags, but also (if both are available) prefer them over the coordinates in the Exif-description tag**, in case somone [edited the geodata before upload](https://github.com/mapillary/mapillary_issues/issues/561#issuecomment-88104637)

**Comment by kaitu** [comment]
related to #73, #249

**Comment by ybkuang** [comment]
Thanks for reporting @de-vries and thanks for connecting the issues @kaitu. Currently we prioritize the data from EXIF description if lat/lon exist in both places. We can do a patch if gps is 0,0 in EXIF description than we use the lat/lon from EXIF.

**Comment by kaitu** [comment]
Uhm, with my setup heavily relying on an external bluetooth GPS, that would work for me (awkward, but it would work) only if you could also make an option in the app to always force gps 0,0 . Otherwise I would have to both stop the connection to the external bluetooth GPS and switch off the internal one to force gps 0,0. But then I would not be able to use the phone for navigation anymore, and I would also lose the ability to check gps reception quality while shooting.

I cannot rely on internal gps, as getting a fix in area without network coverage and heavy obstructions is close to impossible. But using an external gps, currently also means that I have to rewrite coordinates in exif for best results (see #385 for explanation).

In still hope that you will reconsider the gps 0,0 requirement.

**Comment by ybkuang** [comment]
@kaitu @de-vries Thanks for the feedbacks. Currently in the apps, GPS are recorded in both EXIF and EXIF description. And the GPS precision recorded in EXIF is lower than the one in EXIF description. That is the main reason we prefer the GPS in EXIF description over in EXIF.  We will need to investigate a bit more here to find a better solution.

**Comment by HolgerJeromin** [comment]
If someone is filling you the EXIF gps tag and you have better gps data it would be the best to replace the old data. This confuses many users.

**Comment by ybkuang** [comment]
@HolgerJeromin Agree. We just need a way to check whether the EXIF is filled by the users.

**Comment by HolgerJeromin** [comment]
I do not see your point.
If the image is from the app, it should be "the best we had at shot time". If it was changed since than it is better (as the user has changed it by himself).
If the image is not from the app, you have no additional info and have to be happy with the data.

**Comment by de-vries** [comment]
@ybkuang, thanx for clarifying. I missed that explanation in #73. sorry. 

>We just need a way to check whether the EXIF is filled by the users.

1. If the coordinates in the description tag are missing you can safely take the ones in the EXIF. (This was the case with the upload that brought me to opening this issue)
2. If the the difference between the two sets of coords is big enough this should lead to the conclusion that someone edited the tags and the difference is not due to rounding off the coords. (if I understand things right)
3. You could introduce a exif tag, like `EXIF_GPSinfo_Edited=yes`, to indicate that the coords are edited. Not the most elegant or user friendly solution but it could work

**Comment by ybkuang** [comment]
@HolgerJeromin There will be no problem for images not from the app. 

For images from the app, GPS information is in both [EXIF GPS] tag and [EXIF Image Description] tag. We generally prefer the ones from [EXIF Image Description] tag (which is created by the app) due to its precision.  If a user edit only EXIF GPS tag, it is in general difficult for us to detect. One way will be to check the modification time of an image and compare it to the one in Image Description. But we are concerned that it is not so robust.

**Comment by ybkuang** [comment]
@de-vries 
>> If the coordinates in the description tag are missing you can safely take the ones in the EXIF. (This was the case with the upload that brought me to opening this issue)

Yes. We can do a fix like this.

>> If the the difference between the two sets of coords is big enough this should lead to the conclusion that someone edited the tags and the difference is not due to rounding off the coords. (if I understand things right)

This could work. But it might not be reliable enough. 

>> You could introduce a exif tag, like EXIF_GPSinfo_Edited=yes, to indicate that the coords are edited. Not the most elegant or user friendly solution but it could work

This will work. But as you said, it is not super user-friendly. ;)

Will think of it and see if we can come up with a better solution.

**Comment by HolgerJeromin** [comment]
Is the precision an EXIF header can save so low, you have to store the real value at another place? That suprises me... But than your solution is understandable.

**Comment by ybkuang** [comment]
mapillary_issues is no longer an active repository.
Let’s continue discussing features in our forum! We will import this issue as topic to the Mapillary Forum in the coming weeks. We’ll notify here once it’s been done. See you there!
